File: leet_code/JS/linked_list/merge_k_sorted_lists.js
var mergeKLists = function (lists) {
    if (!lists.length) return null;
    if (lists.length === 1) return lists[0];

    let merged = lists[0];
    // start i from 1
    for (let i = 1; i < lists.length; i++) {
        // merge merged with lists[i] and update merged
        merged = mergeTwoLists(merged, lists[i]);
    }

    return merged;
};

var mergeTwoLists = function (l1, l2) {

    if (!l1 || !l2) return l1 || l2;

    let merged = new ListNode();
    let head = merged;
    while (l1 && l2) {
        if (l1.val <= l2.val) {
            merged.next = l1;
            l1 = l1.next;
        } else {
            merged.next = l2;
            l2 = l2.next;
        }
        merged = merged.next;
    }

    if (l1) merged.next = l1;
    if (l2) merged.next = l2;

    return head.next;

};

File: leet_code/python/linked_list/merge_k_sorted_lists.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def mergeKLists(lists):
        if not lists:
            return

        while len(lists) > 1:
            listOne = lists.pop()
            listTwo = lists.pop()
            merged = mergeTwoLists(listOne, listTwo)
            lists.append(merged)
        return lists[0]

def mergeTwoLists(l1, l2):
    dummy = current = ListNode(0)

    while l1 and l2:
        if l1.val <= l2.val:
            current.next = l1
            l1 = l1.next
            current = current.next
        elif l2.val <= l1.val:
            current.next = l2
            l2 = l2.next
            current = current.next

    current.next = l1 or l2

    return dummy.next
